optuna/storages/_redis.py

```diff
@@ -607,3 +643,48 @@ def _check_trial_id(self, trial_id: int) -> None:
 
         if not self._redis.exists(self._key_trial(trial_id)):
             raise KeyError("study_id {} does not exist.".format(trial_id))
+
+    def record_heartbeat(self, trial_id: int) -> None:
+        study_id = self.get_study_id_from_trial_id(trial_id)
+        self._redis.hset(
+            self._key_study_heartbeats(study_id), str(trial_id), pickle.dumps(time.time())
+        )
+
+    def fail_stale_trials(self, study_id: int) -> List[int]:
+        confirmed = []
+        for trial_id in self._get_stale_trial_ids(study_id):
+            if self.set_trial_state(trial_id, TrialState.FAIL):
+                confirmed.append(trial_id)
+
+        return confirmed
+
+    def _get_stale_trial_ids(self, study_id: int) -> List[int]:
+        assert self._heartbeat_interval is not None
+
+        if self._grace_period is None:
+            grace_period = 2 * self._heartbeat_interval
+        else:
+            grace_period = self._grace_period
+
+        current_time = time.time()
```

Comment:
Can you use the current time of DB side, not client side?

Reply:
Thanks, I was not aware of this problem. Good catch!